src/reflect/scala/reflect/io/ZipArchive.scala

```diff
@@ -123,37 +123,54 @@ final class FileZipArchive(file: JFile) extends ZipArchive(file) {
   lazy val (root, allDirs) = {
     val root = new DirEntry("/")
     val dirs = mutable.HashMap[String, DirEntry]("/" -> root)
-    val zipFile = try {
+
+    def openZipFile(): ZipFile = try {
       new ZipFile(file)
     } catch {
       case ioe: IOException => throw new IOException("Error accessing " + file.getPath, ioe)
     }
 
+    val zipFile = openZipFile()
     val enum    = zipFile.entries()
 
-    while (enum.hasMoreElements) {
-      val zipEntry = enum.nextElement
-      val dir = getDir(dirs, zipEntry)
-      if (zipEntry.isDirectory) dir
-      else {
-        class FileEntry() extends Entry(zipEntry.getName) {
-          override def getArchive   = zipFile
-          override def lastModified = zipEntry.getTime()
-          override def input        = getArchive getInputStream zipEntry
-          override def sizeOption   = Some(zipEntry.getSize().toInt)
+    try {
+      while (enum.hasMoreElements) {
+        val zipEntry = enum.nextElement
+        val dir = getDir(dirs, zipEntry)
+        if (zipEntry.isDirectory) dir
+        else {
+          // allows zipEntry to be garbage collected
+          val name = zipEntry.getName()
+          val time = zipEntry.getTime()
+          val size = zipEntry.getSize().toInt
+          class FileEntry() extends Entry(name) {
```

Comment:
I'd suggest moving this class out to an enclosing scope where it can see `openZipFile` but not `zipEntry` and `zipFile`. (Basically, to make the capture explicit as constructor arguments.)